src/internal/task/task_coroutine.go

```diff
@@ -0,0 +1,106 @@
+// +build scheduler.coroutines
+
+package task
+
+import (
+	"unsafe"
+)
+
+// rawState is an underlying coroutine state exposed by llvm.coro.
+// This matches *i8 in LLVM.
+type rawState uint8
+
+//go:export llvm.coro.resume
+func (s *rawState) resume()
+
+//go:export llvm.coro.destroy
+func (s *rawState) destroy()
+
+//go:export llvm.coro.done
+func (s *rawState) done() bool
+
+//go:export llvm.coro.promise
+func (s *rawState) _promise(alignment int32, from bool) unsafe.Pointer
+
+type state struct{ *rawState }
+
+//go:export llvm.coro.noop
+func noopState() *rawState
+
+// Resume the task until it pauses or completes.
+func (t *Task) Resume() {
+	t.state.resume()
+}
+
+// setState is used by the compiler to set the state of the function at the beginning of a function call.
+// Returns the state of the caller.
+func (t *Task) setState(s *rawState) *rawState {
+	caller := t.state
+	t.state = state{s}
+	return caller.rawState
+}
+
+// returnTo is used by the compiler to return to the state of the caller.
+func (t *Task) returnTo(parent *rawState) {
+	t.state = state{parent}
+	t.returnCurrent()
+}
+
+// returnCurrent is used by the compiler to return to the state of the caller in a case where the state is not replaced.
+func (t *Task) returnCurrent() {
+	scheduleTask(t)
+}
+
+//go:linkname scheduleTask runtime.runqueuePushBack
+func scheduleTask(*Task)
+
+// setReturnPtr is used by the compiler to store the return buffer into the task.
+// This buffer is where the return value of a function that is about to be called will be stored.
+func (t *Task) setReturnPtr(buf unsafe.Pointer) {
+	t.Ptr = buf
+}
+
+// getReturnPtr is used by the compiler to get the return buffer stored into the task.
+// This is called at the beginning of an async function, and the return is stored into this buffer immediately before resuming the caller.
+func (t *Task) getReturnPtr() unsafe.Pointer {
+	return t.Ptr
+}
+
+// createTask returns a new task struct initialized with a no-op state.
+func createTask() *Task {
+	return &Task{
+		state: state{noopState()},
+	}
+}
+
+//go:noescape
+
+// start invokes a function in a new goroutine. Calls to this are inserted by the compiler.
+// The created goroutine starts running immediately.
+// This is implemented inside the compiler.
+func start(fn uintptr, args unsafe.Pointer)
+
+// Current returns the current active task.
+// This is implemented inside the compiler.
+func Current() *Task
+
+// Pause suspends the current running task.
+// This is implemented inside the compiler.
+func Pause()
+
+type taskHolder interface {
+	setState(*rawState) *rawState
+	returnTo(*rawState)
+	returnCurrent()
+	setReturnPtr(unsafe.Pointer)
+	getReturnPtr() unsafe.Pointer
+}
+
+var _ = taskHolder((*Task)(nil))
+
+func fake() {
+	// Hack to ensure intrinsics are discovered.
```

Comment:
Would it be possible to add the internal/task package to this list instead?

https://github.com/tinygo-org/tinygo/blob/1d913a62bc0ceef841953058c2c6fb1dd2a40f6a/ir/passes.go#L75

It's an ugly hack, but at least keeping the exceptional packages in one place makes it easier to remove it once it is not necessary anymore.

Reply:
https://github.com/tinygo-org/tinygo/pull/828/files#diff-137ed90530f3e44de5ea2195d32b734eL75 I did. However that does not solve the entire problem, and so I unfortunately need both.

Reply:
Ok, I understand.